typewriter/fixes/tests/test_util.py

```diff
@@ -0,0 +1,114 @@
+
+        passing_tests = ((None, "a", "z", "from x import a as z"),
+                         (None, "a", "z", "import a as z"),
+                         (None, "a.b", "z", "import a.b as z"),
+                         ("a", "b", "z", "import a.b as z"),
+                         ("a.b", "c", "z", "import a.b.c as z"),
+                         (None, "a", "z", "import a as z\nimport u as v"),
+                         (None, "a", "w", "from x import b as y, c as z, a as w, d"),
+                         (None, "a", "z", "from x.b import a as z"),
+                         (None, "a", "z", "from x.b import b as y, c as w, a as z, d"),
+                         (None, "a", "z", "import b as x, c as y, a as z, d as w"))
+        for package, name, alias, import_ in passing_tests:
+            n = self.get_import_binding_info(import_, name, package)
+            self.assertTrue(n)
+            self.assertTrue(n.import_name == name)
+            self.assertTrue(n.binding_name == alias)
+
+    def test_in_function(self):
+        self.try_with("def foo():\n\tbar.baz()\n\tstart=3")
+
+
+class Test_find_import_info(TestCase):
+
+    def find_import_info(self, name, string, package):
+        node = parse(string)
+        print(node)
+        # Return the node as an import binding if possible, None if otherwise
+        return fixer_utils.find_import_info(name, node, package)
+
+    def try_with(self, string):
+        failing_tests = (("a", "a", "x", "from a import b as x"),
+                         ("a.d", "a", "x", "from a.d import b as x"),
+                         ("d.a", "a", "x", "from d.a import b as x"),
+                         (None, "a", "x", "import b as x"),
+                         (None, "a", "x", "import b as x, c, d as y"),
+                         (None, "y", "b", "from x import y as a"))
+        for package, name, alias, import_ in failing_tests:
+            n = self.find_import_info(name, import_ + '\n' + string, package)
+            if n:
+                # If valid binding, check to make sure import info is correct
+                self.assertFalse(n.binding_name == alias and n.import_name == name)
+            else:
+                # Should return None
+                self.assertFalse(n)
+
+        passing_tests = ((None, "a", "z", "from x import a as z"),
+                         (None, "a", "z", "import a as z"),
+                         (None, "a.b", "z", "import a.b as z"),
+                         ("a", "b", "z", "import a.b as z"),
+                         ("a.b", "c", "z", "import a.b.c as z"),
+                         (None, "a", "z", "import a as z\nimport u as v"),
+                         (None, "a", "w", "from x import b as y, c as z, a as w, d"),
+                         (None, "a", "z", "from x.b import a as z"),
+                         (None, "a", "z", "from x.b import b as y, c as w, a as z, d"),
+                         (None, "a", "z", "import b as x, c as y, a as z, d as w"))
```

Comment:
Similar to above, now that we unconditionally check aliases, I want to make sure we always get the right result even with 'similar' imports:

```
(None, "AnotherClass", "bar", """
# Attempt to disguise as bar - before real import
import bar as tofu
import mod3.bar
import mod3.bar as thing
from mod3 import bar as other
from mod3.submod import bar as blah
from mod3.bar import bar as tricksy

# real import
from mod2 import AnotherClass as bar

# Attempt to disguise as bar - after real import
import bar as tofu2
import mod4.bar
import mod4.bar as thing2
from mod4 import bar as other2
from mod4.submod import bar as blah2
from mod4.bar import bar as tricksy2"""),

(None, "mod2.AnotherClass", "bar", """
# Attempt to disguise as mod2.AnotherClass - before real import
import something as AnotherClass
import AnotherClass
import AnotherClass as other
import mod3.AnotherClass
import mod3.AnotherClass as thing
import mod2.stuff as AnotherClass
import mod3.mod2.AnotherClass
import mod3.mod2.AnotherClass as ox
import mod3.mod2.dingo as AnotherClass
from something import geese as AnotherClass
from AnotherClass import goober
from AnotherClass import goober as gobble
from mod3 import AnotherClass
from mod3 import AnotherClass as bucket
from mod2 import stuff as AnotherClass
from mod3.mod2 import AnotherClass
from mod3.mod2 import AnotherClass as trex
from mod3.mod2 import dingo as AnotherClass

# real import
import mod2.AnotherClass as bar

# Attempt to disguise as mod2.AnotherClass - after real import
import something2 as AnotherClass
import AnotherClass
import AnotherClass as other2
import mod4.AnotherClass
import mod4.AnotherClass as thing2
import mod2.stuff2 as AnotherClass
import mod4.mod2.AnotherClass
import mod4.mod2.AnotherClass as ox2
import mod4.mod2.dingo2 as AnotherClass
from something2 import geese2 as AnotherClass
from AnotherClass import goober2
from AnotherClass import goober2 as gobble2
from mod4 import AnotherClass
from mod4 import AnotherClass as bucket2
from mod4 import stuff2 as AnotherClass
from mod4.mod2 import AnotherClass
from mod4.mod2 import AnotherClass as trex2
from mod4.mod2 import dingo2 as AnotherClass"""),
```

Reply:
Oops - looking back on this, I realized that I got confused - the "disguise as bar" test above probably isn't needed - we never "search for" aliases.

Reply:
So, I would like to see the second test - the "Keep this one!" above - added, just to make sure things are working.  (I think it should pass, but it's a good regression test, if nothing else!)

Reply:
I had to tweak it a little. The test shouldn't have no package set. Our new way of querying for data requires us to right hand split the import name if possible, passing in the right item of the split as the import entry and the left split as the package. So I just made the package `mod2` and name `AnotherClass` and it works.